### DISCUSSION TITLE
How can implement one-time task with data in Springboot Application?

### DISCUSSION THREAD

**Discussion by tianshuainan** [kagkarlsson/db-scheduler#377] [Q&A]
I refer to given example to implement one-time task [https://github.com/kagkarlsson/db-scheduler#spring-boot-usage](https://github.com/kagkarlsson/db-scheduler#spring-boot-usage),but still occur some errors.
# Prerequisites

I am using SpringBoot application, and imported dependency 
 ```
  <dependency>
            <groupId>com.github.kagkarlsson</groupId>
            <artifactId>db-scheduler-spring-boot-starter</artifactId>
            <version>11.7</version>
 </dependency>
```

I refer to this link for implementationI [https://github.com/kagkarlsson/db-scheduler#spring-boot-usage](https://github.com/kagkarlsson/db-scheduler#spring-boot-usage) ，and I want to implement a one-time task with data
```
@Service
publuc class myService {

    public static final TaskWithDataDescriptor<CoolingOffTaskData> coolingOffPassedOneTimeTask = new TaskWithDataDescriptor("coolingOffPassedTask-checker", CoolingOffTaskData.class);


   @Autowired
   Scheduler schedulerClient;

    @Bean
    Task<CoolingOffTaskData> sampleOneTimeTask() {
        return Tasks.oneTime(coolingOffPassedOneTimeTask)
                .execute((instance, ctx) -> {
                    final CoolingOffTaskData data = instance.getData();
                    log.info("{} has passed CoolingOff, ready to email", data.getRequestId());
                    });
                });
    }
}
```
I want start one-time task in this method:
  ```
private void startTask(DsrRequest dsrRequest) {
        CoolingOffTaskData coolingOffTaskData = new CoolingOffTaskData(dsrRequest.getId());
        log.info("task_instance COP- {} is saving in scheduled_tasks table", dsrRequest.getId());
        schedulerClient.schedule(coolingOffPassedOneTimeTask.instance("COP-" + dsrRequest.getId(), coolingOffTaskData),
        Instant.now().plusSeconds(150)
        );
```
   
## Current Behavior
Error shows :There is circular references between myService and com.github.kagkarlsson.scheduler.boot.autoconfigure.DbSchedulerAutoConfiguration, that is why and How can I fix it？Can you provide example for my demand？ Thanks！

**Comment by tianshuainan** [comment]
@kagkarlsson Kindly help on this.

**Comment by kagkarlsson** [comment]
Not clear why you have a problem with circular-dependency. That is usually when one wants to use a `SchedulerClient` from within a task or completion-handler. For those use-cases you can use `ctx.getSchedulerClient()` on `ExecutionContext`

**Comment by strelchm** [comment]


**Reply by strelchm** [reply]
```
┌─────┐
|  ratingService
↑     ↓
|  com.github.kagkarlsson.scheduler.boot.autoconfigure.DbSchedulerAutoConfiguration
↑     ↓
|  recalculateRatingJobTask (@Bean from config)
└─────┘
```
In ratingService the SchedulerClient is injected, in recalculateRatingJobTask the ratingService is injected
@kagkarlsson help please, what is wrong. The circular dependency cause is Task, but how can we awoid it

**Comment by kagkarlsson** [comment]
I think maybe either split the RatingService or try to instantiate a standalone client  `SchedulerClient client = SchedulerClient.Builder.create(dataSource).build();`

**Reply by strelchm** [reply]
Thanks, moving client to new configuration fixed problem

**Comment by adamalexandru4** [comment]
I realize this might be a significant change, but I strongly believe schedulerClient should be available for autowiring within task success or failure execution methods, @kagkarlsson.

Currently, the workaround ctx.getSchedulerClient() works for completion handlers, but it’s not accessible in failure handlers.
